### ISSUE TITLE
MA0048: File containing multiple types shouldn't raise warning

### ISSUE THREAD

**Issue by wdolek** [meziantou/Meziantou.Analyzer#428]
**Context**
Rule identifier: _MA0048_
Target Framework: _.NET 6.0_
Language version: _10.0_

**Describe the bug**
When file contains multiple (public) types, [MA0048](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0048.md) doesn't really make sense since file won't be named after type in most cases.

Rule description states:

> When there is only one type in the file, it ensures the type name is the same as the file name ...

**To Reproduce**
Consider following file:

```csharp
// Domain.cs
using System;

public record User(string Name, EmailAddress emailAddress);
public record EmailAddress(string Value);
```

**Proposal / Expected behaviour**
If file contains multiple public types, MA0048 is not applied for any symbol in that file.

**Additional context**
In other languages, e.g. JavaScript, TypeScript, Rust, Go or even F#, it's common that single file contain multiple type declarations. With `record` and `record struct` types which can fit single line, "old" behaviour is becoming obsolete. Having multiple files with single line type declaration suddenly doesn't seem to be appealing nor helpful to understand code-base structure better (subjective ?).

Suppressing rule is simple, imo we should do as few suppressions as possible.

```ini
dotnet_diagnostic.MA0048.severity = none
```

**Comment by meziantou** [comment]
I'm not sure what to do with this issue. When a file contains a single type, the rule makes sense. When you have multiple types, there are multiple options:
- No validation
- The first type must match the file name
- One type must match the file name
- Report all types that don't match the file name, so there is only one type per file

Maybe it should be an option in the editorconfig file:

````
# draft:
MA0048.mode = all_types_must_match_file_name       # A file can only contain 1 type and it must match the file name
MA0048.mode = first_type_must_match_file_name      # Only the first type of a file must match the file name
MA0048.mode = one_type_must_match_file_name        # Only one type, whatever the location, must match the file name
MA0048.mode = do_not_validate_when_multiple_types  # Ignore all files that contain multiple types
````

From an implementation point of view, the rule must be rewritten. You need to have the full file context to know what to do. So, you cannot use `RegisterSymbolAction`. Instead, you need to use the the syntax methods (`RegisterSyntaxNodeAction`). This means the rule may be harder to maintain as you need to keep the code up to date with all type declaration syntaxes. If C# vNext introduces a new way to declare a type, the rule must be updated to support this syntax.

**Comment by wdolek** [comment]
Hi @meziantou, thanks for reply. I'm aware that rule would need to be rewritten as it's currently unaware of context - so this is not simple fix indeed.

For start, perhaps re-wording rule description should be enough

> **When there is only one type in the file**, it ensures the type name is the same as the file name.

---

I agree that default should match behaviour of most projects, apparently my experience differs - and sadly I spend a lot of time tweaking analyzer rules in .editorconfig, thus it doesn't feel like right _default_ to me. Though thanks for tip on sharing styles, this is definitely handy!

**Comment by meziantou** [comment]
> For start, perhaps re-wording rule description should be enough

I'll update the documentation to make it clearer

> and sadly I spend a lot of time tweaking analyzer rules in .editorconfig, thus it doesn't feel like right default to me

Some projects want to be explicit and configure almost all rules manually <https://github.com/dotnet/runtime/blob/main/eng/CodeAnalysis.src.globalconfig>. On my side, I try to limit the number of configured rules, but I still have a few dozen lines in my editorconfig.

**Comment by meziantou** [comment]
I would say it's a similar rule as SA1649 support configuration and seems to exclude partial classes. Do you want to submit a PR to update the documentation page?
